File: src/main/java/cz/cvut/fel/plichjan/distmesh/MeshReader.java
package cz.cvut.fel.plichjan.distmesh;

import cz.cvut.fel.plichjan.distmesh.result.Mesh;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

/**
 */
public class MeshReader extends TriangleFormatReader {
    private List<int[]> triangles;
    private List<double[]> nodes;

    public MeshReader(String name) throws IOException {
        BufferedReader ele = initTokenizer(name + ".ele");
        parseTriangles();
        ele.close();

        BufferedReader node = initTokenizer(name + ".node");
        parseVertices();
        node.close();
    }

    public Mesh getMesh() {
        final Mesh mesh = new Mesh();
        mesh.setP(nodes.toArray(new double[nodes.size()][]));
        mesh.setT(triangles.toArray(new int[triangles.size()][]));
        return mesh;
    }

    private void parseTriangles() throws IOException {
        // parse header
        int nTriangles = readInt();
        int nodePerTriangle = readInt();
        int nAttribute = readInt();
//        readEOL();

        triangles = new ArrayList<int[]>(nTriangles);

        if ((nodePerTriangle != 6 && nodePerTriangle != 3)) {
            throw new IOException();
        }

        for (int i = 0; i < nTriangles; i++) {
            readInt(); // triangle index
            triangles.add(getNodes(nodePerTriangle));
            readIntArr(nAttribute); // triangle attributes
//            readEOL();
        }
    }

    private int[] getNodes(int nodePerTriangle) throws IOException {
        if (nodePerTriangle == 6) {
            int[] srcNodes = readIntArr(6);
            return new int[] {srcNodes[0], srcNodes[1], srcNodes[2], srcNodes[5], srcNodes[3], srcNodes[4]};
        } else {
            int[] srcNodes = readIntArr(3);
            return new int[] {srcNodes[0], srcNodes[1], srcNodes[2]};
        }
    }

    private void parseVertices() throws IOException {
        // parse header
        int nVertex = readInt();
        int dimension = readInt();
        int nAttribute = readInt();
        int boundaryMarker = readInt();
//        readEOL();

        nodes = new ArrayList<double[]>(nVertex);
        if (dimension != 2 || boundaryMarker != 1) {
            throw new IOException();
        }

        for (int i = 0; i < nVertex; i++) {
            readInt(); // vertex index
            double x = readDouble();
            double y = readDouble();
            nodes.add(new double[]{x, y});

            readDoubles(nAttribute);
            readInt(); // boundaryMarker
//            readEOL();
        }
    }

}

File: src/main/kotlin/cz/cvut/fel/plichjan/distmesh/MeshReader.kt
package cz.cvut.fel.plichjan.distmesh

import cz.cvut.fel.plichjan.distmesh.result.Mesh
import java.io.*
import java.util.ArrayList

class MeshReader @Throws(IOException::class) constructor(name: String) : TriangleFormatReader() {
    private var triangles: MutableList<IntArray>? = null
    private var nodes: MutableList<DoubleArray>? = null

    init {
        val ele = initTokenizer("$name.ele")
        parseTriangles()
        ele.close()

        val node = initTokenizer("$name.node")
        parseVertices()
        node.close()
    }

    val mesh: Mesh
        get() {
            val mesh = Mesh()
            mesh.p = nodes!!.toTypedArray()
            mesh.t = triangles!!.toTypedArray()
            return mesh
        }

    @Throws(IOException::class)
    private fun parseTriangles() {
        // parse header
        val nTriangles = readInt()
        val nodePerTriangle = readInt()
        val nAttribute = readInt()

        triangles = ArrayList(nTriangles)

        if (nodePerTriangle != 6 && nodePerTriangle != 3) {
            throw IOException()
        }

        for (i in 0 until nTriangles) {
            readInt() // triangle index
            triangles!!.add(getNodes(nodePerTriangle))
            readIntArr(nAttribute) // triangle attributes
        }
    }

    @Throws(IOException::class)
    private fun getNodes(nodePerTriangle: Int): IntArray {
        return if (nodePerTriangle == 6) {
            val srcNodes = readIntArr(6)
            intArrayOf(srcNodes[0], srcNodes[1], srcNodes[2], srcNodes[5], srcNodes[3], srcNodes[4])
        } else {
            val srcNodes = readIntArr(3)
            intArrayOf(srcNodes[0], srcNodes[1], srcNodes[2])
        }
    }

    @Throws(IOException::class)
    private fun parseVertices() {
        // parse header
        val nVertex = readInt()
        val dimension = readInt()
        val nAttribute = readInt()
        val boundaryMarker = readInt()

        nodes = ArrayList(nVertex)
        if (dimension != 2 || (boundaryMarker != 1 && boundaryMarker != 0)) {
            // Some .node files might have 0 boundary markers
            // throw new IOException();
        }

        for (i in 0 until nVertex) {
            readInt() // vertex index
            val x = readDouble()
            val y = readDouble()
            nodes!!.add(doubleArrayOf(x, y))

            readDoubles(nAttribute)
            readInt() // boundaryMarker
        }
    }
}
